// _ant-design-vue@3.0.0-beta.9@ant-design-vue/es/vc-mentions/src/mentionsProps.js
import _extends from "@babel/runtime/helpers/esm/extends";
import PropTypes from '../../_util/vue-types';
import { initDefaultProps } from '../../_util/props-util';
import { filterOption as defaultFilterOption, validateSearch as defaultValidateSearch } from './util';
import { tuple } from '../../_util/type';
export var PlaceMent = tuple('top', 'bottom');
export var mentionsProps = {
  autofocus: PropTypes.looseBool,
  prefix: PropTypes.oneOfType([PropTypes.string, PropTypes.arrayOf(PropTypes.string)]),
  prefixCls: PropTypes.string,
  value: PropTypes.string,
  disabled: PropTypes.looseBool,
  split: PropTypes.string,
  transitionName: PropTypes.string,
  placement: PropTypes.oneOf(PlaceMent),
  character: PropTypes.any,
  characterRender: PropTypes.func,
  filterOption: {
    type: [Boolean, Function]
  },
  validateSearch: PropTypes.func,
  getPopupContainer: {
    type: Function
  },
  options: {
    type: Array,
    default: function _default() {
      return undefined;
    }
  },
  loading: PropTypes.looseBool,
  rows: [Number, String],
  direction: {
    type: String
  }
};
export var vcMentionsProps = _extends({}, mentionsProps);
export var defaultProps = {
  prefix: '@',
  split: ' ',
  rows: 1,
  validateSearch: defaultValidateSearch,
  filterOption: function filterOption() {
    return defaultFilterOption;
  }
};
export default initDefaultProps(vcMentionsProps, defaultProps);
