// _ant-design-vue@3.0.0-beta.9@ant-design-vue/lib/vc-table/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Column", {
  enumerable: true,
  get: function get() {
    return _Column.default;
  }
});
Object.defineProperty(exports, "ColumnGroup", {
  enumerable: true,
  get: function get() {
    return _ColumnGroup.default;
  }
});
Object.defineProperty(exports, "INTERNAL_COL_DEFINE", {
  enumerable: true,
  get: function get() {
    return _legacyUtil.INTERNAL_COL_DEFINE;
  }
});
Object.defineProperty(exports, "Summary", {
  enumerable: true,
  get: function get() {
    return _Footer.FooterComponents;
  }
});
Object.defineProperty(exports, "SummaryCell", {
  enumerable: true,
  get: function get() {
    return _Footer.SummaryCell;
  }
});
Object.defineProperty(exports, "SummaryRow", {
  enumerable: true,
  get: function get() {
    return _Footer.SummaryRow;
  }
});
exports.default = void 0;

var _Table = _interopRequireDefault(require("./Table"));

var _Footer = require("./Footer");

var _Column = _interopRequireDefault(require("./sugar/Column"));

var _ColumnGroup = _interopRequireDefault(require("./sugar/ColumnGroup"));

var _legacyUtil = require("./utils/legacyUtil");

// base rc-table@7.17.2
var _default = _Table.default;
exports.default = _default;
