// _ant-design-vue@3.0.0-beta.9@ant-design-vue/lib/progress/props.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.progressStatuses = exports.progressProps = void 0;

var _vueTypes = _interopRequireDefault(require("../_util/vue-types"));

var _type = require("../_util/type");

var progressStatuses = (0, _type.tuple)('normal', 'exception', 'active', 'success');
exports.progressStatuses = progressStatuses;
var ProgressType = (0, _type.tuple)('line', 'circle', 'dashboard');
var ProgressSize = (0, _type.tuple)('default', 'small');

var progressProps = function progressProps() {
  return {
    prefixCls: _vueTypes.default.string,
    type: _vueTypes.default.oneOf(ProgressType),
    percent: _vueTypes.default.number,
    format: {
      type: Function
    },
    status: _vueTypes.default.oneOf(progressStatuses),
    showInfo: _vueTypes.default.looseBool,
    strokeWidth: _vueTypes.default.number,
    strokeLinecap: _vueTypes.default.oneOf((0, _type.tuple)('butt', 'round', 'square')),
    strokeColor: {
      type: [String, Object]
    },
    trailColor: _vueTypes.default.string,
    width: _vueTypes.default.number,
    success: {
      type: Object,
      default: function _default() {
        return {};
      }
    },
    gapDegree: _vueTypes.default.number,
    gapPosition: _vueTypes.default.oneOf((0, _type.tuple)('top', 'bottom', 'left', 'right')),
    size: _vueTypes.default.oneOf(ProgressSize),
    steps: _vueTypes.default.number,

    /** @deprecated Use `success` instead */
    successPercent: _vueTypes.default.number,
    title: String
  };
};

exports.progressProps = progressProps;
