// _ant-design-vue@3.0.0-beta.9@ant-design-vue/es/vc-drawer/src/IDrawerPropTypes.js
import _extends from "@babel/runtime/helpers/esm/extends";
import PropTypes from '../../_util/vue-types';

var props = function props() {
  return {
    prefixCls: PropTypes.string,
    width: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    height: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    style: PropTypes.style,
    class: PropTypes.string,
    placement: {
      type: String
    },
    wrapperClassName: PropTypes.string,
    level: {
      type: [String, Array]
    },
    levelMove: {
      type: [Number, Function, Array]
    },
    duration: PropTypes.string,
    ease: PropTypes.string,
    showMask: PropTypes.looseBool,
    maskClosable: PropTypes.looseBool,
    maskStyle: PropTypes.style,
    afterVisibleChange: PropTypes.func,
    keyboard: PropTypes.looseBool,
    contentWrapperStyle: PropTypes.style,
    autofocus: PropTypes.looseBool,
    open: PropTypes.looseBool
  };
};

var drawerProps = function drawerProps() {
  return _extends(_extends({}, props()), {
    forceRender: PropTypes.looseBool,
    getContainer: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object, PropTypes.looseBool])
  });
};

var drawerChildProps = function drawerChildProps() {
  return _extends(_extends({}, props()), {
    getContainer: PropTypes.func,
    getOpenCount: PropTypes.func,
    scrollLocker: PropTypes.any,
    switchScrollingEffect: PropTypes.func
  });
};

export { drawerProps, drawerChildProps };
