// _ant-design-vue@3.0.0-beta.9@ant-design-vue/lib/collapse/commonProps.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.collapseProps = exports.collapsePanelProps = void 0;

var _type = require("../_util/type");

var _vueTypes = _interopRequireDefault(require("../_util/vue-types"));

var collapseProps = function collapseProps() {
  return {
    prefixCls: _vueTypes.default.string,
    activeKey: {
      type: [Array, Number, String]
    },
    defaultActiveKey: {
      type: [Array, Number, String]
    },
    accordion: _vueTypes.default.looseBool,
    destroyInactivePanel: _vueTypes.default.looseBool,
    bordered: _vueTypes.default.looseBool,
    expandIcon: _vueTypes.default.func,
    openAnimation: _vueTypes.default.object,
    expandIconPosition: _vueTypes.default.oneOf((0, _type.tuple)('left', 'right')),
    collapsible: {
      type: String
    },
    ghost: _vueTypes.default.looseBool
  };
};

exports.collapseProps = collapseProps;

var collapsePanelProps = function collapsePanelProps() {
  return {
    openAnimation: _vueTypes.default.object,
    prefixCls: _vueTypes.default.string,
    header: _vueTypes.default.any,
    headerClass: _vueTypes.default.string,
    showArrow: _vueTypes.default.looseBool,
    isActive: _vueTypes.default.looseBool,
    destroyInactivePanel: _vueTypes.default.looseBool,

    /** @deprecated Use `collapsible="disabled"` instead */
    disabled: _vueTypes.default.looseBool,
    accordion: _vueTypes.default.looseBool,
    forceRender: _vueTypes.default.looseBool,
    expandIcon: _vueTypes.default.func,
    extra: _vueTypes.default.any,
    panelKey: _vueTypes.default.oneOfType([_vueTypes.default.string, _vueTypes.default.number]),
    collapsible: {
      type: String
    },
    role: String,
    onItemClick: {
      type: Function
    }
  };
};

exports.collapsePanelProps = collapsePanelProps;
