// _ant-design-vue@3.0.0-beta.9@ant-design-vue/lib/dropdown/props.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.dropdownProps = exports.dropdownButtonProps = exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _type = require("../_util/type");

var _vueTypes = _interopRequireDefault(require("../_util/vue-types"));

var _buttonTypes = _interopRequireDefault(require("../button/buttonTypes"));

var dropdownProps = function dropdownProps() {
  return {
    arrow: _vueTypes.default.looseBool,
    trigger: {
      type: [Array, String]
    },
    overlay: _vueTypes.default.any,
    visible: _vueTypes.default.looseBool,
    disabled: _vueTypes.default.looseBool,
    align: {
      type: Object
    },
    getPopupContainer: _vueTypes.default.func,
    prefixCls: _vueTypes.default.string,
    transitionName: _vueTypes.default.string,
    placement: _vueTypes.default.oneOf((0, _type.tuple)('topLeft', 'topCenter', 'topRight', 'bottomLeft', 'bottomCenter', 'bottomRight')),
    overlayClassName: _vueTypes.default.string,
    overlayStyle: _vueTypes.default.style,
    forceRender: _vueTypes.default.looseBool,
    mouseEnterDelay: _vueTypes.default.number,
    mouseLeaveDelay: _vueTypes.default.number,
    openClassName: _vueTypes.default.string,
    minOverlayWidthMatchTrigger: _vueTypes.default.looseBool,
    onVisibleChange: {
      type: Function
    },
    'onUpdate:visible': {
      type: Function
    }
  };
};

exports.dropdownProps = dropdownProps;
var ButtonTypesProps = (0, _buttonTypes.default)();

var dropdownButtonProps = function dropdownButtonProps() {
  return (0, _extends2.default)((0, _extends2.default)({}, dropdownProps()), {
    type: ButtonTypesProps.type,
    size: _vueTypes.default.oneOf((0, _type.tuple)('small', 'large')),
    htmlType: ButtonTypesProps.htmlType,
    href: _vueTypes.default.string,
    disabled: _vueTypes.default.looseBool,
    prefixCls: _vueTypes.default.string,
    icon: _vueTypes.default.any,
    title: _vueTypes.default.string,
    onClick: {
      type: Function
    }
  });
};

exports.dropdownButtonProps = dropdownButtonProps;
var _default = dropdownProps;
exports.default = _default;
