// _ant-design-vue@3.0.0-beta.9@ant-design-vue/es/progress/props.js
import PropTypes from '../_util/vue-types';
import { tuple } from '../_util/type';
export var progressStatuses = tuple('normal', 'exception', 'active', 'success');
var ProgressType = tuple('line', 'circle', 'dashboard');
var ProgressSize = tuple('default', 'small');
export var progressProps = function progressProps() {
  return {
    prefixCls: PropTypes.string,
    type: PropTypes.oneOf(ProgressType),
    percent: PropTypes.number,
    format: {
      type: Function
    },
    status: PropTypes.oneOf(progressStatuses),
    showInfo: PropTypes.looseBool,
    strokeWidth: PropTypes.number,
    strokeLinecap: PropTypes.oneOf(tuple('butt', 'round', 'square')),
    strokeColor: {
      type: [String, Object]
    },
    trailColor: PropTypes.string,
    width: PropTypes.number,
    success: {
      type: Object,
      default: function _default() {
        return {};
      }
    },
    gapDegree: PropTypes.number,
    gapPosition: PropTypes.oneOf(tuple('top', 'bottom', 'left', 'right')),
    size: PropTypes.oneOf(ProgressSize),
    steps: PropTypes.number,

    /** @deprecated Use `success` instead */
    successPercent: PropTypes.number,
    title: String
  };
};
