// _ant-design-vue@3.0.0-beta.9@ant-design-vue/es/vc-select/hooks/useOptions.js
import { shallowRef, watchEffect } from 'vue';
import { convertChildrenToData } from '../utils/legacyUtil';
/**
 * Parse `children` to `options` if `options` is not provided.
 * Then flatten the `options`.
 */

export default function useOptions(options, children, fieldNames) {
  var mergedOptions = shallowRef();
  var valueOptions = shallowRef();
  var labelOptions = shallowRef();
  watchEffect(function () {
    var newOptions = options.value;
    var childrenAsData = !options.value;

    if (childrenAsData) {
      newOptions = convertChildrenToData(children.value);
    }

    var newValueOptions = new Map();
    var newLabelOptions = new Map();

    function dig(optionList) {
      var isChildren = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;

      // for loop to speed up collection speed
      for (var i = 0; i < optionList.length; i += 1) {
        var option = optionList[i];

        if (!option[fieldNames.value.options] || isChildren) {
          newValueOptions.set(option[fieldNames.value.value], option);
          newLabelOptions.set(option[fieldNames.value.label], option);
        } else {
          dig(option[fieldNames.value.options], true);
        }
      }
    }

    dig(newOptions);
    mergedOptions.value = newOptions;
    valueOptions.value = newValueOptions;
    labelOptions.value = newLabelOptions;
  });
  return {
    options: mergedOptions,
    valueOptions: valueOptions,
    labelOptions: labelOptions
  };
}
