// _ant-design-vue@3.0.0-beta.9@ant-design-vue/lib/vc-drawer/src/IDrawerPropTypes.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.drawerProps = exports.drawerChildProps = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _vueTypes = _interopRequireDefault(require("../../_util/vue-types"));

var props = function props() {
  return {
    prefixCls: _vueTypes.default.string,
    width: _vueTypes.default.oneOfType([_vueTypes.default.string, _vueTypes.default.number]),
    height: _vueTypes.default.oneOfType([_vueTypes.default.string, _vueTypes.default.number]),
    style: _vueTypes.default.style,
    class: _vueTypes.default.string,
    placement: {
      type: String
    },
    wrapperClassName: _vueTypes.default.string,
    level: {
      type: [String, Array]
    },
    levelMove: {
      type: [Number, Function, Array]
    },
    duration: _vueTypes.default.string,
    ease: _vueTypes.default.string,
    showMask: _vueTypes.default.looseBool,
    maskClosable: _vueTypes.default.looseBool,
    maskStyle: _vueTypes.default.style,
    afterVisibleChange: _vueTypes.default.func,
    keyboard: _vueTypes.default.looseBool,
    contentWrapperStyle: _vueTypes.default.style,
    autofocus: _vueTypes.default.looseBool,
    open: _vueTypes.default.looseBool
  };
};

var drawerProps = function drawerProps() {
  return (0, _extends2.default)((0, _extends2.default)({}, props()), {
    forceRender: _vueTypes.default.looseBool,
    getContainer: _vueTypes.default.oneOfType([_vueTypes.default.string, _vueTypes.default.func, _vueTypes.default.object, _vueTypes.default.looseBool])
  });
};

exports.drawerProps = drawerProps;

var drawerChildProps = function drawerChildProps() {
  return (0, _extends2.default)((0, _extends2.default)({}, props()), {
    getContainer: _vueTypes.default.func,
    getOpenCount: _vueTypes.default.func,
    scrollLocker: _vueTypes.default.any,
    switchScrollingEffect: _vueTypes.default.func
  });
};

exports.drawerChildProps = drawerChildProps;
